azure-ipam/ipam_test.go

```diff
@@ -0,0 +1,303 @@
+	default:
+		result := &cns.IPConfigResponse{
+			PodIpInfo: cns.PodIpInfo{
+				PodIPConfig: cns.IPSubnet{
+					IPAddress:    "10.0.1.10",
+					PrefixLength: 24,
+				},
+				NetworkContainerPrimaryIPConfig: cns.IPConfiguration{
+					IPSubnet: cns.IPSubnet{
+						IPAddress:    "10.0.1.0",
+						PrefixLength: 24,
+					},
+					DNSServers:       nil,
+					GatewayIPAddress: "10.0.0.1",
+				},
+				HostPrimaryIPInfo: cns.HostIPInfo{
+					Gateway:   "10.0.0.1",
+					PrimaryIP: "10.0.0.1",
+					Subnet:    "10.0.0.0/24",
+				},
+			},
+			Response: cns.Response{
+				ReturnCode: 0,
+				Message:    "",
+			},
+		}
+		return result, nil
+	}
+}
+
+func (c *MockCNSClient) ReleaseIPAddress(ctx context.Context, ipconfig cns.IPConfigRequest) error {
+	switch ipconfig.InfraContainerID {
+	case "failRequestCNSReleaseIPArgs":
+		return errFoo
+	default:
+		return nil
+	}
+}
+
+const (
+	happyPodArgs    = "K8S_POD_NAMESPACE=testns;K8S_POD_NAME=testname;K8S_POD_INFRA_CONTAINER_ID=testid"
+	nothappyPodArgs = "K8S_POD_NAMESPACE=testns;K8S_POD_NAME=testname;K8S_POD_INFRA_CONTAINER_ID=testid;break=break"
+)
+
+type scenario struct {
+	name    string
+	args    *cniSkel.CmdArgs
+	wantErr bool
+}
+
+func TestCmdAdd(t *testing.T) {
+	req := require.New(t)
+
+	happyNetConf := &cniTypes.NetConf{
+		CNIVersion: "0.1.0",
+		Name:       "happynetconf",
+	}
+
+	invalidVersionNetConf := &cniTypes.NetConf{
+		CNIVersion: "0",
+		Name:       "nothappynetconf",
+	}
+
+	happyNetConfByteArr, err := json.Marshal(happyNetConf)
+	if err != nil {
+		panic(err)
+	}
+	invalidVersionNetConfByteArr, err := json.Marshal(invalidVersionNetConf)
+	if err != nil {
+		panic(err)
+	}
+	invalidNetConf := []byte("invalidNetConf")
+
+	happyArgs := &cniSkel.CmdArgs{
+		ContainerID: "happyArgs",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        happyPodArgs,
+		StdinData:   happyNetConfByteArr,
+	}
+
+	failCreateCNSReqArgs := &cniSkel.CmdArgs{
+		ContainerID: "failCreateCNSReqArgs",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        nothappyPodArgs,
+		StdinData:   happyNetConfByteArr,
+	}
+
+	failRequestCNSArgs := &cniSkel.CmdArgs{
+		ContainerID: "failRequestCNSArgs",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        happyPodArgs,
+		StdinData:   happyNetConfByteArr,
+	}
+
+	failProcessCNSResp := &cniSkel.CmdArgs{
+		ContainerID: "failProcessCNSResp",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        happyPodArgs,
+		StdinData:   happyNetConfByteArr,
+	}
+
+	failParseNetConf := &cniSkel.CmdArgs{
+		ContainerID: "failParseNetConf",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        happyPodArgs,
+		StdinData:   invalidNetConf,
+	}
+
+	failGetVersionedResult := &cniSkel.CmdArgs{
+		ContainerID: "failGetVersionedResult",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        happyPodArgs,
+		StdinData:   invalidVersionNetConfByteArr,
+	}
```

Comment:
do we need to add any scenarios for network plugin = overlay?

Reply:
can you elaborate on this?

Reply:
looping in @thatmattlong